oauthenticator/cilogon.py

```diff
@@ -241,144 +273,105 @@ def _validate_allowed_idps(self, proposal):
+        primary username_claim.
         """,
-        default_value=["email"],
     )
 
-    def _get_final_username_claim_list(self, user_info):
+    def user_info_to_username(self, user_info):
         """
-        The username claims that will be used to determine the hub username can be set through:
-         - `CILogonOAutnenticator.username_claim`, that can be extended through `CILogonOAutnenticator.additional_username_claims`
-         or
-         - `CILogonOAuthenticator.allowed_idps.<idp>.username_claim`, that
-            will overwrite any value set through CILogonOAuthenticator.username_claim
-            for this identity provider.
-
-        This function returns the username claim list that will be used for the current user trying to login
-        based on the idp that they have selected. If no `CILogonOAutnenticator.allowed_idps` is set, then
-        `CILogonOAutnenticator.username_claim` will be used.
+        Overrides OAuthenticator.user_info_to_username that relies on
+        username_claim to instead consider idp specific config in under
+        allowed_idps[user_info["idp"]]["username_derivation"].
+
+        Returns a username based on user_info and configuration in allowed_idps
+        under the associated idp's username_derivation config.
         """
-        username_claims = [self.username_claim]
-        if self.additional_username_claims:
-            username_claims.extend(self.additional_username_claims)
-        if self.allowed_idps:
-            selected_idp = user_info["idp"]
-            if selected_idp in self.allowed_idps.keys():
-                # The username_claim which should be used for this idp
-                return [
-                    self.allowed_idps[selected_idp]["username_derivation"][
-                        "username_claim"
-                    ]
-                ]
-            else:
-                return username_claims
-        return username_claims
-
-    def _get_username_from_claim_list(self, user_info, username_claims):
-        username = None
-        for claim in username_claims:
-            username = user_info.get(claim)
-            if username:
-                break
+        # NOTE: The first time we have received user_info is when
+        #       user_info_to_username is called by OAuthenticator.authenticate,
+        #       so we make a check here that the "idp" claim is received and
+        #       that we allowed_idps is configured to handle that idp.
+        #
+        user_idp = user_info.get("idp")
+        if not user_idp:
+            message = "'idp' claim was not part of the response to the userdata_url"
+            self.log.error(message)
+            raise web.HTTPError(500, message)
+        if not self.allowed_idps.get(user_idp):
+            message = f"Login with identity provider {user_idp} is not pre-configured"
+            self.log.error(message)
+            raise web.HTTPError(500, message)
```

Comment:
Personally, I think a [403 status code](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/403) might be more suited here, since the idp is not **alowed** to login, not just not configured.
Also, since the idps made visible through cilogon's `shown_idps` feature now come from the actual `allowed_idps`, someone seeing this error would need to edit the url and remove the shown_idps query parameters to be presented the entire idp list.

Reply:
Deal, updated in 8d1da50!